### ISSUE TITLE
AWS has changed Elasticache Redis versioning again

### ISSUE THREAD

**Issue by philomory** [hashicorp/terraform-provider-aws#22385]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

```
Terraform v1.0.7
on windows_amd64
+ provider registry.terraform.io/hashicorp/aws v3.70.0
```

### Affected Resource(s)

* `aws_elasticache_replication_group`

(This probably also impacts `aws_elasticache_cluster` when used with Redis, but I haven't personally verified it)

### Terraform Configuration Files

Please include all Terraform configurations required to reproduce the bug. Bug reports without a functional reproduction may be closed without investigation.

```hcl
terraform {
  backend "local" { }

  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 3.70"
    }
  }
}

provider "aws" {
  region  = "us-east-1"
}

resource "aws_elasticache_replication_group" "example1" {
  replication_group_id          = "tf-rep-group-1"
  replication_group_description = "test description"
  engine_version                = "6.0"
  node_type                     = "cache.m4.large"
  parameter_group_name          = "default.redis6.x"
}

resource "aws_elasticache_replication_group" "example2" {
  replication_group_id          = "tf-rep-group-2"
  replication_group_description = "test description"
  engine_version                = "6.2"
  node_type                     = "cache.m4.large"
  parameter_group_name          = "default.redis6.x"
}
```

### Debug Output

https://gist.github.com/philomory/eeead53871f26c6a3825e36c17a862af

### Expected Behavior

Create two Redis clusters, one at 6.0 and one at 6.2

### Actual Behavior

An error is displayed, stating `engine_version: Redis versions must match <major>.x when using version 6 or higher, or <major>.<minor>.<bug-fix>`.

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

Previously, Amazon had announced the following for Redis versions 6 and higher:
> Beginning with Redis 6, ElastiCache for Redis will offer a single version for each Redis OSS major release, rather than offering multiple minor versions. Versionless engine support is designed to minimize confusion and ambiguity on having to choose from multiple minor versions. ElastiCache for Redis will also automatically keep your cache cluster up to date on the selected major version (Redis 6 and above). By providing the updated engine version, it ensures improved performance and enhanced security.
> 
> You specify the engine version by using 6.x. ElastiCache for Redis will automatically invoke the preferred minor version of Redis 6 that is available.

(This was as opposed to with versions 5.x and lower, where you could specify full version numbers).

However, at some point in the recent past (_probably_ coinciding with the [2021-11-23 announcement of support for Redis 6.2](https://aws.amazon.com/about-aws/whats-new/2021/11/amazon-elasticache-redis-support-redis-6-2/), but I'm not certain), the documentation was changed to state:
> Beginning with Redis 6.0, ElastiCache for Redis will offer a single version for each Redis OSS minor release, rather than offering multiple patch versions.
> 
> Beginning with Redis 6.0, you can also opt-in to the next auto minor version upgrade by setting the AutoMinorVersionUpgrade parameter to yes, and ElastiCache for Redis will manage the minor version upgrade, through self-service updates. This will be handled through standard customer-notification channels via a self-service update campaign. For more information, see Self-service updates in Amazon ElastiCache.

The output of `aws elasticache describe-cache-engine-versions` now confirms that there are engine versions `"6.0"` and `"6.2"` now, but _not_ any engine version identified as `"6.x"` (although oddly attempting to create a cluster with engine_version "6.x" _does_ still work, and results in a cluster using version 6.2.5).

### References

* [Elasticache Redis 6.2 support announcement](https://aws.amazon.com/about-aws/whats-new/2021/11/amazon-elasticache-redis-support-redis-6-2/)
* [Elasticache Redis supported engine versions](https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/supported-engine-versions.html)
* [Elasticache Redis uprading versions](https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/VersionManagement.html)
* #15625, the issue initially resulting in implementation of support for `6.x` as an `engine_version` and _removal_ of support for e.g. `6.0`.
* #18920, the pull request that resulted from the above-mentioned issue

**Comment by manelpb** [comment]
Yes Please!!! This is basically blocking us to upgrade to version 6.0

**Comment by ktham** [comment]
So what needs to change?

Does that Elasticache API now want us to specify `6.0.x` and `6.2.x` rather than `6.x` for `engine_version` now for **creating** an elasticache cluster? 

> The output of aws elasticache describe-cache-engine-versions now confirms that there are engine versions "6.0" and "6.2" now, but not any engine version identified as "6.x" (although oddly attempting to create a cluster with engine_version "6.x" does still work, and results in a cluster using version 6.2.5).

That has always been the case. The Describe API always return the **actual** version number that is running (i.e. 5.0.5, 6.0, etc...), not necessarily the engine_version parameter specified in the create API call. (i.e. "6.x").

**Comment by ktham** [comment]
I've been looking at the docs, indeed, it seems like with the 6.2 release, it seems like AWS wants us to use "6.0" and "6.2" in the `engine_version` field 🤦  I'm guessing AWS is still supporting "6.x" for now but silently wants us to migrate with that documentation change.

**Comment by ktham** [comment]
> Yes Please!!! This is basically blocking us to upgrade to version 6.0

@manelpb - Seems like you can still specify "6.x" in the `engine_version` field and still be fine and create the latest available 6.x cluster (i.e. 6.2.5). That was the advertised behavior for a long time now. (i.e. not being able to specify the minor version and create the latest 6.x one).

**Comment by manelpb** [comment]
> > Yes Please!!! This is basically blocking us to upgrade to version 6.0
> 
> @manelpb - Seems like you can still specify "6.x" in the `engine_version` field and still be fine and create the latest available 6.x cluster (i.e. 6.2.5). That was the advertised behavior for a long time now. (i.e. not being able to specify the minor version and create the latest 6.x one).

Yes, but that would force us to migrate to 6.2 instead of 6.0. Even though they don't have any significant breaking change, there was indeed a change that could affect the way they calculate key expiration, as described here: 

```
Command behavior changes:
* EXISTS should not alter LRU (#8016)
  In Redis 5.0 and 6.0 it would have touched the LRU/LFU of the key.
```
https://raw.githubusercontent.com/redis/redis/6.2/00-RELEASENOTES

We don't want to migrate right from 5.0 to 6.2 without making sure it won't affect all of our servers, which might take some time.

**Comment by ktham** [comment]
> @ktham Yes, but that would force us to migrate to 6.2 instead of 6.0. Even though they don't have any significant breaking change, there was indeed a change that could affect the way they calculate key expiration

Ah, I see. cc @gdavison, there might be some tweaks needed to https://github.com/hashicorp/terraform-provider-aws/pull/18920 to support setting `engine_version` to `"6.2"` and `"6.0"` in addition to `"6.x"` now that AWS wants us to specify the minor version in addition to the major version.

I think we'll want to:
* Update the regular expression used for validation here to allow "MAJOR.MINOR" where minor can be an "x" **or** a numeric value: https://github.com/hashicorp/terraform-provider-aws/blob/2d0b6c10802bd0edae5c8f58b6a35489a0d86e3e/internal/service/elasticache/validation.go#L17
* And update the wording here: https://github.com/hashicorp/terraform-provider-aws/blob/2d0b6c10802bd0edae5c8f58b6a35489a0d86e3e/internal/service/elasticache/validation.go#L36

**Comment by philomory** [comment]
> > The output of aws elasticache describe-cache-engine-versions now confirms that there are engine versions "6.0" and "6.2" now, but not any engine version identified as "6.x" (although oddly attempting to create a cluster with engine_version "6.x" does still work, and results in a cluster using version 6.2.5).
> 
> That has always been the case for the 6.x clusters. The Describe API always return the **actual** version number that is running (i.e. 5.0.5, 6.0.0, etc...), not necessarily the engine_version parameter specified in the create API call. (i.e. "6.x").

Note that I'm not talking about the API call used to describe the version that a particular cluster is running, this API call returns a list of versions (for both Redis and Memcached) that are supported by the API. I'm pretty sure you're thinking of `describe-cache-clusters`, not `describe-cache-engine-versions`.

**Comment by ktham** [comment]
>  I'm pretty sure you're thinking of describe-cache-clusters, not describe-cache-engine-versions.

Ah, indeed. Sorry, I didn't read your PR description carefully initially

**Comment by MatiasNielsen** [comment]
Using engine_version "6.x" seems to work.

**Comment by KMahankali** [comment]
Do we have any ETA when will be the provider this will be available to support 6.0 and 6.2 for engine_version?

**Comment by bschaatsbergen** [comment]
> Do we have any ETA when will be the provider this will be available to support 6.0 and 6.2 for engine_version?

I'll high likely get the PR ready today, up to the core contributors, as ElastiCache is a core service you might get likely and it becomes part of 4.6.0 (upcoming release).

**Comment by manelpb** [comment]
Thank you @bschaatsbergen

**Comment by bschaatsbergen** [comment]
I had to jump between a few hoops in order to make this work as friendly as possible without breaking anything. I removed the `6.x` support as upgrading your resource from `6.0` to `6.x` is not possible, the ModifyCacheCluster API call seems to ignore it and your cluster will stay on `6.0.5` instead of upgrading. The regex pattern now no longer accepts `6.x`, it only accepts <major>.<minor>, meaning you're in charge of either supplying `6.0` or `6.2` for example. 

Thoughts?

**Comment by manelpb** [comment]
As long we understand that this would cause a breaking change and we add a note about it, I think it should be fine

**Comment by bschaatsbergen** [comment]
Hey all, the fix is merged and will be part of 4.12.0 🎉

**Comment by manelpb** [comment]
> Hey all, as you can see by the above the fix is merged and will be part of 4.12.0 🎉

Thank you so much!!
